cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -310,124 +300,91 @@ func generateMountsModeRecover(mst *initramfsMountsState, recoverySystem string)
-	isMounted, err := mst.IsMounted(boot.InitramfsUbuntuSeedDir)
+	err := mountPartitionMatchingKernelDisk(boot.InitramfsUbuntuSeedDir, "ubuntu-seed")
 	if err != nil {
-		return false, err
-	}
-	if !isMounted {
-		return false, selectPartitionMatchingKernelDisk(boot.InitramfsUbuntuSeedDir, "ubuntu-seed")
+		return err
 	}
 
 	// 2.1. measure model
 	err = stampedAction(fmt.Sprintf("%s-model-measured", recoverySystem), func() error {
 		return secbootMeasureSnapModelWhenPossible(mst.Model)
 	})
 	if err != nil {
-		return false, err
+		return err
 	}
 
 	// 2.2. (auto) select recovery system and mount seed snaps
-	isBaseMounted, err := mst.IsMounted(filepath.Join(boot.InitramfsRunMntDir, "base"))
-	if err != nil {
-		return false, err
-	}
-	isKernelMounted, err := mst.IsMounted(filepath.Join(boot.InitramfsRunMntDir, "kernel"))
+	typs := []snap.Type{snap.TypeBase, snap.TypeKernel, snap.TypeSnapd}
+	essSnaps, err := mst.RecoverySystemEssentialSnaps("", typs)
 	if err != nil {
-		return false, err
+		return fmt.Errorf("cannot load metadata and verify essential bootstrap snaps %v: %v", typs, err)
 	}
-	isSnapdMounted, err := mst.IsMounted(filepath.Join(boot.InitramfsRunMntDir, "snapd"))
-	if err != nil {
-		return false, err
-	}
-	if !isBaseMounted || !isKernelMounted || !isSnapdMounted {
-		// load the recovery system and generate mounts for kernel/base
-		// and snapd
-		var whichTypes []snap.Type
-		if !isBaseMounted {
-			whichTypes = append(whichTypes, snap.TypeBase)
-		}
-		if !isKernelMounted {
-			whichTypes = append(whichTypes, snap.TypeKernel)
-		}
-		if !isSnapdMounted {
-			whichTypes = append(whichTypes, snap.TypeSnapd)
-		}
-		essSnaps, err := mst.RecoverySystemEssentialSnaps("", whichTypes)
-		if err != nil {
-			return false, fmt.Errorf("cannot load metadata and verify essential bootstrap snaps %v: %v", whichTypes, err)
-		}
 
-		// TODO:UC20: do we need more cross checks here?
-		for _, essentialSnap := range essSnaps {
-			switch essentialSnap.EssentialType {
-			case snap.TypeBase:
-				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, "base"))
-			case snap.TypeKernel:
-				// TODO:UC20: we need to cross-check the kernel path with snapd_recovery_kernel used by grub
-				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, "kernel"))
-			case snap.TypeSnapd:
-				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, "snapd"))
-			}
+	// TODO:UC20: do we need more cross checks here?
+	for _, essentialSnap := range essSnaps {
+		dir := snapTypeToMountDir[essentialSnap.EssentialType]
+		// TODO:UC20: we need to cross-check the kernel path with snapd_recovery_kernel used by grub
+		err := doSystemdMount(essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, dir), nil)
+		if err != nil {
+			return err
 		}
 	}
 
+	// TODO:UC20: after we have the kernel and base snaps mounted, we should do
```

Comment:
Well that is already handled in the initrd right now with https://github.com/snapcore/core-initrd/blob/main/factory/usr/lib/systemd/system/sysroot-usr-lib-firmware.mount and https://github.com/snapcore/core-initrd/blob/main/factory/usr/lib/systemd/system/sysroot-usr-lib-modules.mount, So yes we can merge this PR without implementing this TODO, since this TODO is just about migrating from those systemd units to snap-bootstrap proper in Go code. But once this PR is merged, it should be easy to add those mounts to snap-bootstrap here and eliminate them from the initrd fs layout.

But to answer your question anyways, if we didn't do those mounts in snap-bootstrap and we didn't have those mount units we would not have kernel modules and thus would not have networking and probably a lot of other things too, so we should only drop those mounts from the initrd after snap-bootstrap is doing them.

To keep things stable for those mount units while making this transition, we may need to make them conditional on the mounts not already being there otherwise they may fail if snap-bootstrap does the mounts and the mount units are still there. I need to investigate this more, but again makes more sense as a followup instead of doing it here.

Reply:
The other related thing I want to get rid of from the initrd fs layout is https://github.com/snapcore/core-initrd/blob/main/factory/usr/lib/systemd/system/sysroot-writable.mount and bits of the ExecStart from https://github.com/snapcore/core-initrd/blob/main/factory/usr/lib/systemd/system/populate-writable.service which call https://github.com/snapcore/core-initrd/blob/main/factory/usr/lib/the-modeenv